### PR TITLE
In "induction", add a clause "over" more explicit than "in" for generalizing the induction hypothesis over hypotheses of the context

### PR DESCRIPTION
**Pull request by herbelin**
The main outcome of this PR is to provide a notation `induction n over m` as a more explicit and lighter syntax than `induction n in m |- *` to generalize the induction hypothesis of an induction.

**Kind:** feature

- [X] Corresponding documentation was added / updated.
- [x] Entry added in ~~CHANGES~~ the changelog

Typical example:
```coq
Require Import Arith.
Goal forall m n, n + m = m + n.
induction n over m.
- now rewrite Nat.add_0_l.
- now rewrite Nat.add_succ_l, IHn.
Qed.
```

Side questions: I'm unsure of the English. Does `over` make sense in this context or another preposition should better be used?

### CODE REVIEW

**Comment by maximedenes** [comment]
I'm not sure, but I thought "induction over m" would mean what "induction m" does.

Who can review this PR?

**Comment by Zimmi48** [comment]
It could be also `induction n generalizing m`. Input from a native English speaker is definitely required here.

**Comment by maximedenes** [comment]
That looks heavier than the current syntax...

**Comment by Zimmi48** [comment]
> That looks heavier than the current syntax...

The `generalizing` alternative? Yes indeed, though it's not the only thing to consider. Other criteria may include for instance the ease to remember the syntax. Obviously power users will want to a shorter syntax.

If I understand correctly, this new syntax is equivalent (at least in the example above) to `intros until n; revert m; induction n; intro m.` So power ssreflect users would do something like `move=> n m; move: m; elim.` and even better `move=> + m; elim.` with the new intro patterns from #6705.

But for beginners, `induction n over m.` or `induction n generalizing m.` might be nicer to use than `revert m; induction n; ...` or `induction n in m |- *`.

**Comment by JasonGross** [comment]
As a native English speaker, `induction n over m` seems fine, `induction n generalizing m` is so many characters that I'll never type it ( but it makes sense when I read it), `induction n in m` feels inconsistent with other `in` clauses, and `induction n in m |- *` seems mostly fine.

**Comment by herbelin** [comment]
What is the conclusion of this PR? Is the `induction term over hyps` syntax strictly clearer than the `induction term in hyps |- *`? (At least, that `|- *` is not needed looks less hackish to me.)

Shall I try a poll among more users (and/or beginners) to get more feedback?

Note that the PR includes also a few side improvements (improving context in with induction is interpreted, highlighting). So, a possiblity is also that I cut the PR into more elementary ones.

@JasonGross: would you like e.g. to be the assignee on this PR (i.e. coordinating the discussion and merging if this is the decision).

**Comment by Zimmi48** [comment]
Yes, splitting in more logical units would make sense.
Yes, I think that this new syntax is clearer for beginners. That being said, we should come up with methods to measure this more objectively.
I don't think that the new merging process requires the PR author to propose an assignee, especially when this person is not among the maintainers of the parts which are affected by the PR.

**Comment by Zimmi48** [comment]
There can be many relevant people to conduct reviews. However, the persons who are entitled to merging a PR are the maintainers (GitHub code owners). In this case, @ppedrot should self-assign this PR as the maintainer of `tactics/` which is AFAICS the main component affected. That he is already doing a lot is not the question here, but of course if *he* feels like we are asking him too much, we could reconsider some of the attributions. See also https://github.com/coq/coq/blob/master/dev/doc/MERGING.md#code-owners.

**Comment by herbelin** [comment]
OK, turning then to @ppedrot: "when maintainers receive a review request, they are expected to put their name in the assignee field, if they are in charge of the component that is the main target of the patch".

**Review by forestjulien — APPROVED** [review]

**Review by ppedrot — APPROVED** [review]
No particular opinion on this PR.

**Review by ejgallego — COMMENTED** [review]
This adds yet another element to a gigantic tuple in induction_destruct, but the main problem I see here is that generalization/cleaning should not be implemented by each individual tactic, as it doesn't scale and seems to lead to code duplication / inconsistencies.

**Comment by herbelin** [comment]
@mattam82: Good questions. Here are a few answers:

- One can mix `in` and `over`, but t is an error if an hypothesis is mentioned twice. 
- One can have the effect of `over` by using the `in` clause (at least for compatibility), but not conversely: `over` is intended for hypotheses/variables with no dependency in the term matched and it therefore does not support selecting occurrences of the term matched.
- One other difference between `over` and `in` is that `in` tells the exact list of hypotheses and occurrences on which to generalize, clearing other dependent hypotheses, while `over` only tells what to add to the induction hypothesis which is not already detected to be dependent. If both are present, `over` adds its list to the one of `over` and dependent hyps which are not mentioned are cleared. For instance:
  ```coq
  Goal forall n p : nat, n + p = 0 -> n = 0 -> n = 0.
  intros * H H0.
  induction n in H |- * over p. (* generalizes over p and H and clears H0 *)
  ```
- A `in` or `over` at the end is only for compatibility to support the syntax `induction n using foo in H |- *`. It works only with a single argument.
- A distinct `in` clause can be given for each argument of a multi-`destruct` or multi-`induction`.

I now changed the PR so that `over` is an error for `destruct` (and thus an error for the after-the-first arguments of `induction` since `induction` behaves as `destruct` from its second argument). I now also made two other additions to refine the behavior:
- `over` idents are now space separated so that there is no confusion in `induction n over H, p` with `p` taken as second argument of `induction` and not second hypothesis for `over`;
- I added a warning when `over` is used on an hypothesis which is already generalized by default.

Two further questions:
- Order of variables in the `over` and `in` currently have no impact. Maybe should it matter, so that the order given in `over` and `in` determines the order in which the induction hypothesis is generalized?
- In the following situation:
  ```coq
  Parameter P : nat -> Prop.
  Lemma L : forall n p, P n -> P p -> P (n+p).
  intros * Hn Hp.
  induction n over p.
  ```
  is it more intuitive that `Hp` is automatically generalized because of the required generalization over `p` or is it more intuitive that an explicit `induction n over p Hp` has to be given? I.e., should `over x y` means "generalizing over x, y and their dependencies", or only "generalizing over x and y"?

**Comment by herbelin** [comment]
@ppedrot: I locally made an overlay for Ltac2 but I'm finally a bit bothered. Shall we change the binding of an internal tactic in Ltac2 everytime the API of an internal tactic changes? Shouldn't we try instead, for the purpose of a tactic language, to take some time to think at a stable selection of low-level combinators to export rather than exporting high-level evolving tactics like `induction`. Or, if ever you export `induction` in Ltac2, to only export the most basic versions, e.g. without multiple arguments, without `in`.

**Comment by herbelin** [comment]
@Zimmi48 
> Moreover, the overlay needs to additionally test if the branch is pr-6836 to be picked up by GitLab CI (see #7505).

I'm unsure I understand your comment. Is the overlay that I wrote correct (by imitating other existing overlays) or does it lack a test for GitLab?

**Comment by Zimmi48** [comment]
> I'm unsure I understand your comment. Is the overlay that I wrote correct (by imitating other existing overlays) or does it lack a test for GitLab?

No need to bother since #7505 has been merged. However, you didn't rebase on a recent master, otherwise you wouldn't see all these CircleCI tests. There is also a very weird error in that the build fails only on Travis (which tests the merge of the PR and current master) so it seems to indicate an improper interaction (semantic conflict).

I suggest that you rebase again to solve this.

**Comment by Zimmi48** [comment]
Dune says:

```
File "pretyping/detyping.mli", line 19, characters 5-13:
Error: Unbound module Tactypes
```

**Comment by ejgallego** [comment]
> File "pretyping/detyping.mli", line 19, characters 5-13: Error: Unbound module Tactypes

It lives in `proofs/` which is above of pretyping. I guess @ppedrot can say more about the intended location.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: Since you are modifying this line, could you transform it like this:

```
.. tacv:: induction @term with @bindings_list as @disj_conj_intro_pattern using @term with @bindings_list in @goal_occurrences over {+ @ident}
          einduction @term with @bindings_list as @disj_conj_intro_pattern using @term with @bindings_list in @goal_occurrences over {+ @ident}
```
This is the form to use when two objects share a common description.
Could you also put some `{?` `}` delimiters around each of the optional clauses? Thanks!

**Review by Zimmi48 — APPROVED** [review]

**Comment by ppedrot** [comment]
@herbelin What is blocking this PR? Is it supposed to ready modulo rebase? Am I right that nobody is opposing the PR otherwise?

Regarding Ltac2 changes of API (sorry for the delay) we should try to provide preserve the API as much as possible, but for now the library is not stable enough to repel this kind of change.

**Comment by herbelin** [comment]
> @herbelin What is blocking this PR? Is it supposed to ready modulo rebase? Am I right that nobody is opposing the PR otherwise?

I think I wanted to work a bit more on the documentation and do some extra checks.

File: theories/ssr/ssreflect.v
Comment (CohenCyril) [inline theories/ssr/ssreflect.v:545]: Please do not reserve a new keyword :cry:

Comment (herbelin) [reply theories/ssr/ssreflect.v:545]: À qui le dis-tu !

In some sense, this keyword is reserved for 2 years already...

More seriously, I agree with avoiding keywords which are necessary only in a well-delimited syntactic context. With a small parser hack, it should be possible to make it only locally a keyword. I'll try to do that.

By the way, I guess you would then agree that we try to avoid having `is` and `of` being keywords?

Comment (CohenCyril) [reply theories/ssr/ssreflect.v:545]: I agree there should be as few keywords as possible by default. Still, there is a difference between adding new keywords in a plugin and adding new pervasive keywords. The addition of ssr `is`, `of` and `over` did not trigger a renaming in existing code / tactic AFAIK.

**Comment by CohenCyril** [comment]
> As a native English speaker, `induction n over m` seems fine

Even though I am not a native speaker, allow me to be skeptical about this phrasing, ... in math papers and textbooks I am used to read, the only occurrences of "induction over" are of the form "induction over a number / set / cardinal / dimension", and they never mean "generalizing", (mostly because the pen and paper mathematical practice usually does not specify what to generalize, except if the strengthening is a non-trivial property). Hence, from my math background point of view, saying "over" instead of "generalizing" is extremely confusing.

File: theories/ssr/ssreflect.v
Comment (herbelin) [reply theories/ssr/ssreflect.v:545]: @CohenCyril: As said, this PR is opened for 2 years, so the knowledge of a possible use of "over" was already there. Anyway, I'm willing to find a way to avoid a new keyword and, at the same time why not (to try) to prevent `is` and `of` to be keywords (people who load libraries requiring mathcomp sometimes complain that this changes their parser unexpectedly, by the way).

Actually, more interestingly and challenging in my opinion, why not to think at either move `of` already on the Coq side (if enough people like it), or to think at whether it could be merged with `` ` ``? And similarly for `if is`.

**Comment by herbelin** [comment]
> Hence, from my math background point of view, saying "over" instead of "generalizing" is extremely confusing.

The problem is that "generalizing" is long. So, couldn't "over" be interpreted also as an ellipsis of "generalizing over"?

**Comment by gares** [comment]
If the annotation (about the part of the context to generalize) is not part of the usual "natural language", then I don't see why it should be phrased in Coq using natural language. It could be rather expressed as a technical annotation. Eg, `induction n gen:m` (after `eqn:H`, but I'm not saying I like this syntax more than another one, just that we do already have syntax that is not natural language inspired).

**Comment by CohenCyril** [comment]
> The problem is that "generalizing" is long. So, couldn't "over" be interpreted also as an ellipsis of "generalizing over"?

I first read this PR today, and the first question that popped in my mind is, in `induction n over m`, is the induction over `n` or `m`... so my answer to your question is: not for me, it cannot be interpreted as an ellipsis, because "over" already has a meaning.

**Comment by herbelin** [comment]
> So my answer to your question is: not for me, it cannot be interpreted as an ellipsis, because "over" already has a meaning in this exact context.

Cannot you read it as `induction on n over m`? With the idea that `on` is on the static object, and `over` is like about moving over the `======` sign?

Or maybe `induction n spanning m`, shorter than `generalizing`.

Otherwise, I find it a bit less fluid, especially for teaching, but, indeed, `gen:n`, even if repeated, as in `gen:n gen:p`, looks like a pretty good alternative, consistent with `eqn:`. We should extend the poll.

Relatively to the question of keywords, a possibility is to hack the function recognizing idents (`Tok.match_pattern`) so that it dynamically looks into a table of local keywords to tell if this an ident allowed in the corresponding local context (the drawback is that we cannot use the name `over` as a variable name in an `induction` invocation). This would go in pair with a special side-effect parsing incantation `add_local_keyword_over` similar to those look-aheads we like. Is this an acceptable hack? (@ppedrot, would you accept it?)

Then, the same could be done for `is` and `of`. @gares, @cohencyril, would you be ready to do it?

More generally, about thinking at how valuable and feasible it would be to unite `of` and `` ` ``, and/or about moving `if is` mainstream, @gares, @CohenCyril, I guess you are the most appropriate persons to talk about it, right?

**Comment by Matafou** [comment]
+1 I also think “over” is misleading.
Alternative propositions:

induction n generalizing H,H’.
induction n including H,H’.
induction n for H,H’.
induction n for H,H’|-.

I like the last one because it means « perform an induction for the goal
H,H’|-thegoal ».

P

Le jeu. 10 sept. 2020 à 13:28, Cyril Cohen <notifications@github.com> a
écrit :

**Comment by herbelin** [comment]
> +1 I also think “over” is misleading. 

> Alternative propositions:
>
> induction n generalizing H,H’. 
>
> induction n including H,H’.
>
> induction n for H,H’. 
>
> induction n for H,H’|-.

The latter is close to `induction n in H, H' |- *` whose effect is also to generalize but while this PR automatically generalized over hypotheses dependent on `H` [1], the `in` variant clears the hypotheses dependent on `H` which are not explicitly listed.

I wonder whether the design of `induction n in H, H' |- *` should not be reworked after all so that it implements the semantics of this PR (i.e. [1]). On the other side, the use of `in` for such semantics seems to be confusing, so, to fit with the standard usage of `in` it is probably better to keep it the semantics of selecting occurrences.

Anyway, if I summarize:

- `induction n generalizing m`: clear but long
- `induction n including m`: (don't know what to think about)
- `induction n over m`: short but confusing
- `induction n for m`: short
- `induction n gen:m`: short, clear, maybe a bit hackish (not natural language) and less immediate to remember but consistent with `eqn:`

**Comment by Matafou** [comment]
I was not aware of this syntax. I guess ideally we should unify it with your PR one way or another. Maybe with a tag on hypothesis of which we want to generalize dependencies.

```coq
induction n in `H1, H2 |- *.
```
means generalize `H1` and its dependencies, `H2` without. Replacing "in" by "for" or "gen" is another question.

The "gen:" tag looks also OK to me.

**Comment by Zimmi48** [comment]
@JasonGross do you want the ping from @coqbot to be refined to only occur, e.g. when there is no test-suite failure, similarly to what we should do for automatic extraction of test cases?

**Comment by ppedrot** [comment]
This PR has stalled. It's trying to add a redundant syntax on the good old criterion of "intuitiveness" which is a provably correct guarantee that it will derail into heavy bikeshedding. It unescapably did. What do we do with it? (Personally, I'd advocate closing.)

**Comment by herbelin** [comment]
As far as I perceive it, this PR (as other PRs about tactics) is a victim of difficulties to take decisions.

Personally, I'm fine with `over` and I'd like this PR to go. This is an incremental step but a compatible, and imho, intuitive one.

On the other side, is there someone thinking that the status quo would be better? (@ppedrot?)

(More generally, what about starting a working group on tactics with who is interested, with the objective of converging rather than diverging? A typical first objective could precisely be to work at specifying and stabilizing a semantics for all `destruct`/`induction`/`elim`/`case` variants.)

**Comment by Zimmi48** [comment]
> (More generally, what about starting a working group on tactics with who is interested, with the objective of converging rather than diverging? A typical first objective could precisely be to work at specifying and stabilizing a semantics for all `destruct`/`induction`/`elim`/`case` variants.)

Yes! I'm in. @JasonGross and @jfehrle might also be interested.

**Comment by ppedrot** [comment]
> On the other side, is there someone thinking that the status quo would be better?

Given that:
- Nobody seems to agree on a syntax because it's a subjective matter of taste
- The syntax is redundant
- Every syntax addition made available unconditionally to all of our users (and especially keyword addition) should be very, very seriously discussed to the point that the burden of the proof should be reversed (let me chant my mantra: *syntax is not modular*)
- This is *adding* syntax non-uniformity instead of removing it

I am indeed quite in disfavour of merging this PR. Also, I don't think you can call that *status quo*, which refers to a stable equilibrium between group positions. Here we have about as many positions as people that commented in the thread, so this is more like maximal isotropy.

**Comment by herbelin** [comment]
>     * Nobody seems to agree on a syntax because it's a subjective matter of taste

I disagree. There is of course subjectivity in syntax but there is also some form of aesthetism which can be shared between people. Moreover, subjectivity does not imply that no decision can eventually be made.

> I am indeed quite in disfavour of merging this PR. Also, I don't think you can call that _status quo_, which refers to a stable equilibrium between group positions. Here we have about as many positions as people that commented in the thread, so this is more like maximal isotropy.

I don't understand your position. Can you give me the name of someone who is satisfied with what you call the current "stable equilibrium". I'd bet that even you is unsatisfied.

I don't see why positions would have to be inherently incompatible. It is a question of willing. Do we want to converge to something that satisfies us more or not? Would _you_ like that we converge to something that satisfies us more or not? 

> This is _adding_ syntax non-uniformity instead of removing it.

I don't understand this either. In what sense is it more non-uniform than `as` or `with`? Do you consider the use of `in` for generalization more uniform? 

On the other side, propositions to remove non-uniformity would be most welcome, but I cannot read in your mind, so you have to tell what your propositions are?

>     * The syntax is redundant

If you go this way, there are plenty of redundant syntax. In the balance, I prefer an intuitive syntax even partially redundant than a syntax difficult to explain. Note in passing that `over x` is not the same as `in x`: it generalizes over the dependencies of `x`, which is I believe what you naturally expect, while with `in`, you have to list also the dependencies.

Are you saying that in the balance, the redundancy is not worth the increased intuitiveness?

>     * Every syntax addition made available unconditionally to all of our users (and especially keyword addition) should be very, very seriously discussed to the point that the burden of the proof should be reversed 

There is no keyword addition in this PR. The word `over` (or `gen:`) arrives in a position which does not require it to be a keyword. This syntax extension is purely local.

> let me chant my mantra: _syntax is not modular_

Is your view that the syntax of tactics should be, say, "lispian", i.e. with a minimal or at least fixed number of specific parsing rules? Would you like we remove the special syntaxes for `intros`, `destruct`, `induction`, etc.? What not after all? Actually, my dream is to be able to provide a 3-level syntax of tactics: a standard intuitive syntax revolving around basic intuitive keywords such as `case`, `induction`, `apply`, an intensive-user syntax based on symbolic keywords like `[= ...]`, `//=`, etc. and a pseudo-natural language for rendering, with a one-to-one correspondence between the three levels. We could also have a 4th level based on a fixed "programming-style" syntax, and that would actually be an excellent idea.

In any case, let us allow ourselves to go ahead. I don't see the point of keeping an unsatisfactory situations for years. Let's go in the direction of our dreams. It is not by blocking a move which is consistent with the current approach to syntax that we will arrive to the situation that you seem to dream. On the contrary, I'd be very happy to support your view and to understand how to make them eventually concrete, but for that, it is on your own wishes that we must focus, not on blocking what is right to others.

**Comment by jfehrle** [comment]
> (More generally, what about starting a working group on tactics with who is interested, with the objective of converging rather than diverging? A typical first objective could precisely be to work at specifying and stabilizing a semantics for all destruct/induction/elim/case variants.)

I'd be happy to participate in discussions and to update the documentation to reflect the stablized (clearer) semantics in the documentation.  However, I think we should be conservative and careful in making incompatible syntax changes.

**Comment by herbelin** [comment]
@jfehrle, @Zimmi48: what about meeting online about induction tactics in the week 5-9 July, something like 6pm or 9am Paris time? Or would it be too late wrt the ongoing work on #14179?

**Comment by herbelin** [comment]
@ppedrot: one of the reasons I like the `over` idea is that a similar idea can be developed for `match`.

As of now, the standard approach to typing `match` is to have dependency only in the conclusion:
```
Γ ⊢ t : I u    Γ, y : A, x : I y ⊢ P : Type    Γ, z : B ⊢ v : P[y:=w][x:=C z] 
------------------------------------------------------------------------------
Γ ⊢ (match t as x in I y return P with C z => v end) : P[y:=u][x:=t]
```
and to use a generalizing "commutative cut" when rewriting in an hypothesis of `Γ` is also needed, using expressions of the form:
```
match t as x in I y return forall d:D, P with
| C z => fun d : D[y:=w][x:=C z] => v
end d
```
for `d : D[y:=u][x:=t]`.

Currently, besides introducing a "commutative cut", such generalization can roughly also be simulated with a syntax of the form `match t, x with C z, x => v end` (telling to generalize over the dependencies of `t` and `u` in `x`), but this syntax is a bit tricky and non-intuitive.

So, my idea is to instead provide an abstraction level hard-wiring the ability to rewrite in an hypothesis, as in:
```
Γ ⊢ t : I u   Γ, y : A, x : I y,  Γ' ⊢ P : Type   Γ, z : B, Γ'[y:=w][x:=C z] ⊢ v : P[y:=w][x:=C z]
---------------------------------------------------------------------------------------------------
Γ, Γ'[y:=u][x:=t] ⊢ (match t as x in I y over Γ' return P with C z => v end) : P[y:=u][x:=t]
```
In addition to being quite intuitive, this level of abstraction follows a principle of independence of connectives, that is, that `match` is able to rewrite in the context without having to assume the existence of dependent function type in the ambient logic.

Then, my point is that there is a nice story to tell with all of `match`, `destruct`, `induction` uniformly providing the explicit ability to work on the context, with `over` arriving as a natural keyword to express the abstraction level supporting this ability.

**Comment by herbelin** [comment]
@ppedrot:

Is that what you call modular syntax and syntax uniformity?
```
From Ltac2 Require Import Ltac2.  

Ltac2 g x := Std.destruct false
   [{Std.indcl_arg := Std.ElimOnIdent x;
     Std.indcl_eqn := None;
     Std.indcl_as := None;
     Std.indcl_in := None}].
```

If yes, it seems to suffer more from scalability than an ad-hoc syntax with optional parameters that can be added incrementality without breaking preexisting syntax (as in the case of `over`). So, I'm lost in what you exactly have in mind.

**Comment by jfehrle** [comment]
The week of July 5 would be fine.  6 PM Paris time is 9 AM for me, I expect that's what you meant.  

It would be a mistake to hold up #14179, that's already quite a lot of changes.  If/when the code changes from our discussion, we will update the doc with another PR.  We wouldn't update it before the code changes.

**Comment by Zimmi48** [comment]
Sounds good to me too.

**Comment by herbelin** [comment]
Then, say on Monday the 5th, 6pm?

**Comment by herbelin** [comment]
Let's go on https://rdv2.rendez-vous.renater.fr/coq-call...

**Pull request closed without merging**
